docs/adr/0006_component_constructors.md

```diff
@@ -0,0 +1,363 @@
+      name: internal
+    resources:
+      - name: testdata
+        type: blob
+        relation: local
+        input:
+          type: file
+          path: ./testdata/text.txt
+```
+
+which could be processed by a constructor library as follows:
+
+```yaml
+component:
+  componentReferences: []
+  name: github.com/acme.org/helloworld
+  provider: internal
+  repositoryContexts: []
+  resources:
+  - access:
+      localReference: sha256:c3ab8ff13720e8ad9047dd39466b3c8974e592c2fa383d4a3960714caef0c4f2
+      mediaType: application/octet-stream
+      type: localBlob/v1
+    creationTime: "2025-05-19T08:28:30+02:00"
+    digest:
+      hashAlgorithm: SHA-256
+      normalisationAlgorithm: genericBlobDigest/v1
+      value: c3ab8ff13720e8ad9047dd39466b3c8974e592c2fa383d4a3960714caef0c4f2
+    name: testdata
+    relation: local
+    size: 6
+    type: blob
+    version: 1.0.0
+  sources: []
+  version: 1.0.0
+meta:
+  schemaVersion: v2
+
+``` 
+
+In this example:
+
+- The input of type `file` tells the constructor to use the registered `file` input method.
+- This method processes `./testdata/text.txt` and generates an access specification.
+- The result is included in the final component descriptor.
+- The format of the generated access specification depends on the specific input method used.
+
+---
+
+### Categories of Input Methods
+
+Input methods are typically categorized into two types:
+
+#### Local Input Methods
+
+- Store the content as a `localBlob` next to the generated component version.
+
+#### Global Input Methods
+
+- Generate access specifications that are independent of the constructed component version.
+
+Each input method must determine whether it should produce a `localBlob` or a global access specification.
+
+---
+
+### "By Value" Construction
+
+Any input or access specification can be marked as **constructed by value**. In this case:
+
+- The constructor directly embeds the resource into the component version.
+- The resource is uploaded to the OCM repository as a co-located `localBlob`.
+
+To support this behavior, the constructor must be capable of:
+
+- Uploading the constructed component version.
+- Downloading input data as needed.
+- Uploading resources again as local blobs for inclusion.
+
+## Processing Architecture
+
+All `component-constructors` follow a simple and consistent processing pattern:
+
+1. **Define Specification**   
+   The user specifies the desired component version in a `component-constructor.yaml` file, typically via the CLI.
+
+2. **Parse Component Versions**  
+   The constructor library parses the specified component versions from the YAML file.
+
+3. **Process Resources and Sources**  
+   For each component version, the constructor handles all resources and sources:
+
+    - **Input Method Specified**  
+      If an input method is defined:
+        - If it returns `ResourceInputMethodResult.ProcessedBlobData`, the constructor uploads the local blob to the target OCM repository.
```

Comment:
So this is also upload like in `by value` but the difference here is that it will not add it to the component descriptor?

Reply:
Actually if ProcessedBlobData is provided, then the component version will also have a local blob reference like in "by value". However "by value" forces results that were not some local blob to become local blobs.